ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,241 @@
+   +----------------------------------------------------------------------+
+*/
+
+#include "php.h"
+#include "php_intl.h"
+#include <unicode/ulistformatter.h>
+#include "listformatter_arginfo.h"
+#include "listformatter_class.h"
+#include "intl_convert.h"
+
+static zend_object_handlers listformatter_handlers;
+
+static void listformatter_free_obj(zend_object *object)
+{
+    ListFormatter_object *obj = php_intl_listformatter_fetch_object(object);
+
+    if( obj->lf_data.ulistfmt )
+        ulistfmt_close( obj->lf_data.ulistfmt );
+
+    obj->lf_data.ulistfmt = NULL;
+    intl_error_reset( &obj->lf_data.error );
+
+    zend_object_std_dtor(&obj->zo);
+}
+
+static zend_object *listformatter_create_object(zend_class_entry *class_type)
+{
+    ListFormatter_object *obj;
+    obj = zend_object_alloc(sizeof(ListFormatter_object), class_type);
+
+    obj->lf_data.ulistfmt = NULL;
+    intl_error_reset( &obj->lf_data.error );
+
+    zend_object_std_init(&obj->zo, class_type);
+    object_properties_init(&obj->zo, class_type);
+    obj->zo.handlers = &listformatter_handlers;
+    return &obj->zo;
+}
+
+PHP_METHOD(IntlListFormatter, __construct)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    char* locale;
+    size_t locale_len = 0;
+    zend_long type = ULISTFMT_TYPE_AND;
+    zend_long width = ULISTFMT_WIDTH_WIDE;
+    ZEND_PARSE_PARAMETERS_START(1, 3)
+        Z_PARAM_STRING(locale, locale_len)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG(type)
+        Z_PARAM_LONG(width)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if(locale_len == 0) {
+        locale = (char *)intl_locale_get_default();
+    }
+
+    if (locale_len > INTL_MAX_LOCALE_LEN) {
+        zend_argument_value_error(1, "Locale string too long, should be no longer than %d characters", INTL_MAX_LOCALE_LEN);
+        RETURN_THROWS();
+    }
+
+    if (strlen(uloc_getISO3Language(locale)) == 0) {
+        zend_argument_value_error(1, "\"%s\" is invalid", locale);
+        RETURN_THROWS();
+    }
+
+    UErrorCode status = U_ZERO_ERROR;
+    #if U_ICU_VERSION_MAJOR_NUM >= 67
+        if (type != ULISTFMT_TYPE_AND && type != ULISTFMT_TYPE_OR && type != ULISTFMT_TYPE_UNITS) {
+            zend_argument_value_error(2, "must be one of IntlListFormatter::TYPE_AND, IntlListFormatter::TYPE_OR, or IntlListFormatter::TYPE_UNITS");
+            RETURN_THROWS();
+        }
+        
+        if (width != ULISTFMT_WIDTH_WIDE && width != ULISTFMT_WIDTH_SHORT && width != ULISTFMT_WIDTH_NARROW) {
+            zend_argument_value_error(3, "must be one of IntlListFormatter::WIDTH_WIDE, IntlListFormatter::WIDTH_SHORT, or IntlListFormatter::WIDTH_NARROW");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_openForType(locale, type, width, &status);
+    #else
+        if (type != ULISTFMT_TYPE_AND) {
+            zend_argument_value_error(2, "ICU 66 and below only support IntlListFormatter::TYPE_AND");
```

Comment:
This and the below exception message don't match the beginning of the message generated by `zend_argument_value_error` (Argument #2 ...)

Reply:
Nice catch! Fixed.